From 0d8ac18c61ef6cfc4dacea4f223d8a5d881edbde Mon Sep 17 00:00:00 2001
From: Kailang Yang <kailang@realtek.com>
Date: Thu, 16 May 2019 16:10:44 +0800
Subject: [PATCH] FROMGIT: ALSA: hda/realtek - Check headset type by unplug and resume

BugLink: https://bugs.launchpad.net/bugs/1826181

When system enable HDA power save mode.
In Dell headset mode, it will recheck during runtime resume when headset was plugged.
This patch will move check headset type on unplug and system resume.

Signed-off-by: Kailang Yang <kailang@realtek.com>
(cherry picked from commit df06e42c0153e720ef413e42ba12e8f58e71e358
git://github.com/thesofproject/linux.git)
Signed-off-by: Hui Wang <hui.wang@canonical.com>
---
 sound/pci/hda/patch_realtek.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/sound/pci/hda/patch_realtek.c b/sound/pci/hda/patch_realtek.c
index 0915d7cb9ab2..5673efd6d5df 100644
--- a/sound/pci/hda/patch_realtek.c
+++ b/sound/pci/hda/patch_realtek.c
@@ -4929,6 +4929,8 @@ static void alc_update_headset_mode(struct hda_codec *codec)
 	switch (new_headset_mode) {
 	case ALC_HEADSET_MODE_UNPLUGGED:
 		alc_headset_mode_unplugged(codec);
+		spec->current_headset_mode = ALC_HEADSET_MODE_UNKNOWN;
+		spec->current_headset_type = ALC_HEADSET_TYPE_UNKNOWN;
 		spec->gen.hp_jack_present = false;
 		break;
 	case ALC_HEADSET_MODE_HEADSET:
@@ -4971,8 +4973,6 @@ static void alc_update_headset_mode_hook(struct hda_codec *codec,
 static void alc_update_headset_jack_cb(struct hda_codec *codec,
 				       struct hda_jack_callback *jack)
 {
-	struct alc_spec *spec = codec->spec;
-	spec->current_headset_type = ALC_HEADSET_TYPE_UNKNOWN;
 	snd_hda_gen_hp_automute(codec, jack);
 }
 
@@ -5009,7 +5009,11 @@ static void alc_fixup_headset_mode(struct hda_codec *codec,
 		alc_probe_headset_mode(codec);
 		break;
 	case HDA_FIXUP_ACT_INIT:
-		spec->current_headset_mode = 0;
+		if (codec->core.dev.power.power_state.event == PM_EVENT_RESUME ||
+				codec->core.dev.power.power_state.event == PM_EVENT_RESTORE) {
+			spec->current_headset_mode = 0;
+			spec->current_headset_type = 0;
+		}
 		alc_update_headset_mode(codec);
 		break;
 	}
-- 
2.17.1

